### PR TITLE
[Modules] Added module for Redis Cache Enterprise and Database

### PR DESCRIPTION
**Pull request by ahmadabdalla**
# Description

Created a new module for Redis Cache Enterprise and Database child resource.
Added 3 test cases to validate parameter combinations:
- Common: Deploys a Cluster with Private endpoints and cache persistence
- Min: Minimum requirements.
- Geo Replication: To support an enterprise cluster with Geo replication scenarios.

## Common

image

## Min

image

## Geo replication

image

### Pipeline references

image
image
image
image

| Pipeline |
| - |
| New released Pipeline, tested via Azure DevOps |

## Type of Change

- [ ] New feature (non-breaking change which adds functionality)

## Checklist

- [ ] I'm sure there are no other open Pull Requests for the same update/change
- [ ] My corresponding pipelines / checks run clean and green without any errors or warnings
- [ ] My code follows the style guidelines of this project
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation (readme)
- [ ] I did format my code

### CODE REVIEW

File: modules/cache/redis-enterprise/.test/common/main.test.bicep [hunk 1]
```diff
@@ -0,0 +1,124 @@
...
+        }
+      }
+    ]
+    databases: [
+      {
+        clusteringPolicy: 'EnterpriseCluster'
+        evictionPolicy: 'AllKeysLFU'
+        modules: [
+          {
+            name: 'RedisBloom'
```
Comment (AlexanderSehr) [inline modules/cache/redis-enterprise/.test/common/main.test.bicep:107]: Would you think it makes sense to simplify this format to a flat array and format the object for the user? Only makes sense to me if we don't anticipate additional properties...

File: modules/cache/redis-enterprise/main.bicep [hunk 2]
```diff
@@ -0,0 +1,232 @@
...
+
+@description('Optional. The name of logs that will be streamed. "allLogs" includes all possible logs for the resource. Set to \'\' to disable log collection.')
+@allowed([
+  ''
+  'allLogs'
+  'ConnectionEvents'
+  'audit'
+])
+param diagnosticLogCategoriesToEnable array = [
+  ''
```
Comment (AlexanderSehr) [inline modules/cache/redis-enterprise/main.bicep:80]: Not allLogs here?

Comment (ahmadabdalla) [reply modules/cache/redis-enterprise/main.bicep:80]: Not supported using the standard method. Was getting failures for it. New audit logging ingestion is being introduced in preview

File: modules/cache/redis-enterprise/.test/common/main.test.bicep (hunk 1 above)
Comment (ahmadabdalla) [reply modules/cache/redis-enterprise/.test/common/main.test.bicep:107]: Based on the spec, there's an additional object that can be inserted here:

modules: [
      {
        args: 'string'
        name: 'string'
      }
    ]

Was getting provisioning failures on the settings defined in the doco, without clear message why.

https://learn.microsoft.com/en-us/azure/templates/microsoft.cache/redisenterprise/databases?pivots=deployment-language-bicep

But don't think we should suppress it. Thoughts?

File: modules/cache/redis-enterprise/main.bicep (hunk 2 above)
Comment (AlexanderSehr) [reply modules/cache/redis-enterprise/main.bicep:80]: Hm ok. Should it then even be an option? Maybe we can comment it out and add a disclaimer as to why

File: modules/cache/redis-enterprise/.test/common/main.test.bicep (hunk 1 above)
Comment (AlexanderSehr) [reply modules/cache/redis-enterprise/.test/common/main.test.bicep:107]: @ahmadabdalla If it's supported I guess we should allow it, even if we can't get it to work.

Comment (ahmadabdalla) [reply modules/cache/redis-enterprise/.test/common/main.test.bicep:107]: I'm updating to add a parameter usage section too. Let me do some digging to try to get a working example.

Comment (ahmadabdalla) [reply modules/cache/redis-enterprise/.test/common/main.test.bicep:107]: @AlexanderSehr finally got it to work :) At least now its tested

image

**Review by AlexanderSehr — APPROVED** [review]

**Pull request merged**